CLOUDP-237245: Add example for retrying requests

wtrocki opened the pull request:
## Description

Adding example that would support 
Tested manually  by using API mocks (thanks to @lantoli) for different network issues and status codes. 
All works end to end. 

## Testing
```
go run ./examples/retry/retry.go      
2024/03/26 20:43:41 [DEBUG] GET https://cloud-dev.mongodb.com/api/atlas/v2/groups?includeCount=true&itemsPerPage=1&pageNum=1
2024/03/26 20:43:41 [DEBUG] GET https://cloud-dev.mongodb.com/api/atlas/v2/groups?includeCount=true&itemsPerPage=1&pageNum=1
2024/03/26 20:43:41 [DEBUG] GET https://cloud-dev.mongodb.com/api/atlas/v2/groups?includeCount=true&itemsPerPage=1&pageNum=1
2024/03/26 20:43:42 [DEBUG] GET https://cloud-dev.mongodb.com/api/atlas/v2/groups?includeCount=true&itemsPerPage=1&pageNum=1
```

gssbzn commented on go.mod:
```diff
@@ -7,6 +7,11 @@ require (
 	github.com/mongodb-forks/digest v1.0.5
 )
 
+require (
+	github.com/hashicorp/go-cleanhttp v0.5.2 // indirect
+	github.com/hashicorp/go-retryablehttp v0.7.5 // indirect
+)
```
can this be avoided? hashicorp libs are painful because they are MPL and get flagged in snyk and then it requires legal to approve

wtrocki replied:
yes. I will create dedicated go.mod file for examples in this PR.

wtrocki replied:
Done

wtrocki commented:
[12f3972](https://github.com/mongodb/atlas-sdk-go/pull/294/commits/12f3972d9b04ba6b028100908347f3d3502864a0) was added to not got pinged on dependabot updates for examples. No strong opinion if we want that and we can easily remove this commit if needed.

gssbzn commented on examples/go.mod:
```diff
@@ -0,0 +1,19 @@
+module go.mongodb.org/atlas-sdk/v20231115008/examples
+
+go 1.20
```
latest support go is 
```suggestion
go 1.21
```
if you go above 1.20 you now have to define a full version like 1.21.1

gssbzn commented on examples/go.mod:
```diff
@@ -0,0 +1,19 @@
+go 1.20
+
+require (
+	github.com/go-test/deep v1.1.0
+	github.com/mongodb-forks/digest v1.1.0
+    go.mongodb.org/atlas-sdk/v20231115008 v20231115008.1.0
+    github.com/hashicorp/go-retryablehttp v0.7.5 
```
seems like a format error, please be sure to `go mod tidy`

gssbzn commented on examples/retry/retry.go:
```diff
@@ -0,0 +1,70 @@
+
+}
+
+func createRetryableClient(retryClient *retryablehttp.Client, apiKey string, apiSecret string) *http.Client {
```
```suggestion
func newRetryableClient(retryClient *retryablehttp.Client, apiKey string, apiSecret string) *http.Client {
```

gssbzn commented on examples/retry/retry.go:
```diff
@@ -0,0 +1,70 @@
+	var transport http.RoundTripper = &retryablehttp.RoundTripper{Client: retryClient}
+	digestRetryAbleTransport := digest.NewTransportWithHTTPRoundTripper(apiKey, apiSecret,transport)
+	retryableClient, err := digestRetryAbleTransport.Client()
+	if err != nil {
+		log.Fatal("Cannot instantiate client")
+	}
```
it's ok to panic on main but don't panic on a function return the error and handle it

gssbzn commented on examples/retry/retry.go:
```diff
@@ -0,0 +1,70 @@
+
+func createRetryableClient(retryClient *retryablehttp.Client, apiKey string, apiSecret string) *http.Client {
+	var transport http.RoundTripper = &retryablehttp.RoundTripper{Client: retryClient}
+	digestRetryAbleTransport := digest.NewTransportWithHTTPRoundTripper(apiKey, apiSecret,transport)
```
seems like this file is not `go fmt`
```suggestion
	digestRetryAbleTransport := digest.NewTransportWithHTTPRoundTripper(apiKey, apiSecret, transport)
```

wtrocki replied:
We do have checks for format and run format specifically. Strange that formatter did not fixed it.

gssbzn replied:
you don't https://github.com/mongodb/atlas-sdk-go/blob/b3496fc261e121f206eb9e1479a3fd50b1fbe3ea/.golangci.yml#L90

wtrocki replied:
We do it as part of the [generation](https://github.com/mongodb/atlas-sdk-go/blob/main/tools/scripts/generate.sh#L40) thus ensuring generated is formated, but examples aren't covered there. 

We cannot use gofmt because it would require rewrite of every mustache template we use and will not alow us to update openapi generator.

gssbzn replied:
>not alow us to update openapi generator.

do we plan to ever update? we are already quite behind

wtrocki replied:
I did number of attempts: https://github.com/mongodb/atlas-sdk-go/pull/200
It will mean some breaking changes in the SDK.

gssbzn commented on internal/main.go:
what's this internal main for?

wtrocki commented:
@gssbzn Based on your feedback have:
- Added multi mod support in to the repository
- Moved example dependencies to example project
- Extended this PR scope to update golang version of repository

gssbzn commented on examples/README.md:
```diff
@@ -20,3 +19,11 @@ One of the SDK examples covers usage of the mockery within tests.
+
+Example provides automatic retries for all HTTP 500, 429 HTTP status errors.
+
+`go run ./retry/retry.go`
```
```suggestion
```bash
go run ./retry/retry.go
```
```

gssbzn commented on examples/go.mod:
```diff
@@ -0,0 +1,19 @@
+go 1.22.1
+
+require (
+	github.com/go-test/deep v1.1.0
+	github.com/mongodb-forks/digest v1.1.0
+    go.mongodb.org/atlas-sdk/v20231115008 v20231115008.1.0 
+	github.com/hashicorp/go-retryablehttp v0.7.5 
```
```suggestion
	github.com/go-test/deep v1.1.0
	github.com/mongodb-forks/digest v1.1.0
	go.mongodb.org/atlas-sdk/v20231115008 v20231115008.1.0 
	github.com/hashicorp/go-retryablehttp v0.7.5 
```

wtrocki replied:
go mod tidy do not cover changes for some reasons.

gssbzn commented on go.mod:
is testify still needed here, could we `go mod tidy`?

wtrocki replied:
We do: 

> 	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/mock"

gssbzn commented on examples/retry/retry.go:
```diff
@@ -0,0 +1,76 @@
+	sdk, err := admin.NewClient(
+		admin.UseHTTPClient(retryableClient),
+		admin.UseBaseURL(url),
+		admin.UseDebug(false))
+		if err != nil{
+			log.Fatal(err)
+		}
```
```suggestion
		admin.UseDebug(false))
	if err != nil{
		log.Fatal(err)
	}
```

gssbzn commented on examples/retry/retry.go:
```diff
@@ -0,0 +1,76 @@
+		log.Fatal(err)
+	}
+
+	if projects.GetTotalCount() == 0 {
+		log.Fatal("account should have at least single project")
+	}
```
you can can zero projects so let's make this safer
```suggestion
	log.Info("Total Projects", projects.GetTotalCount())
```

gssbzn commented on examples/retry/retry.go:
```diff
@@ -0,0 +1,76 @@
+func newRetryableClient(retryClient *retryablehttp.Client, apiKey string, apiSecret string) (*http.Client, error) {
+	var transport http.RoundTripper = &retryablehttp.RoundTripper{Client: retryClient}
+	digestRetryAbleTransport := digest.NewTransportWithHTTPRoundTripper(apiKey, apiSecret,transport)
+	retryableClient, err := digestRetryAbleTransport.Client()
+	if err != nil {
+		return nil, err;
+	}
+	return retryableClient, nil
```
```suggestion
	return digestRetryAbleTransport.Client()
```

wtrocki commented:
I would prefer if we do OpenAPI generator update in the follow up PR as this might involve +60 different file changes.

gssbzn reviewed (approved):
LGTM

lantoli commented on examples/README.md:
```diff
@@ -20,3 +19,13 @@ One of the SDK examples covers usage of the mockery within tests.
+
+### Retry Example
+
+Example provides automatic retries for all HTTP 500, 429 HTTP status errors.
```
@wtrocki @gssbzn note that for instance cluster creation can throw 500 but still created the cluster so next call will be DUPLICATE_CLUSTER_NAME 400 that will be a bit confusing for clients.
retries are mostly useful when operations are atomic

gssbzn replied:
yes, personally blindly retrying on the Atlas API is not safe and a bad idea

lantoli commented on examples/go.mod:
```diff
@@ -0,0 +1,20 @@
+
+go 1.22.1
+
+replace go.mongodb.org/atlas-sdk/v20231115008 => ../
```
is this intended to be merged?

wtrocki replied:
Yes
